src/main/scala/com/github/gvolpe/fs2/Counter.scala

```diff
@@ -22,30 +17,31 @@ object CounterApp extends Counter[IO]
   * #2 >> 3
   * #3 >> 3
   * */
-class Counter[F[_] : Effect] extends StreamApp[F] {
-
-  override def stream(args: List[String], requestShutdown: F[Unit]): fs2.Stream[F, ExitCode] =
-    Scheduler(corePoolSize = 10).flatMap { implicit S =>
-      for {
-        ref <- Stream.eval(async.refOf[F, Int](0))
-        w1  = new Worker[F](1, ref)
-        w2  = new Worker[F](2, ref)
-        w3  = new Worker[F](3, ref)
-        ec  <- Stream(w1.start, w2.start, w3.start).join(3).drain ++ Stream.emit(ExitCode.Success)
-      } yield ec
-    }
+object CounterApp extends IOApp {
+  import cats.syntax.all._
+
+  def stream[F[_]: ConcurrentEffect]: Stream[F, Unit] =
+    for {
+      ref <- Stream.eval(Ref.of[F, Int](0))
+      w1 = new Worker[F](1, ref)
+      w2 = new Worker[F](2, ref)
+      w3 = new Worker[F](3, ref)
+      _ <- Stream(w1.start, w2.start, w3.start).parJoin(3)
+    } yield ()
 
+  override def run(args: List[String]): IO[ExitCode] =
+    stream[IO].compile.drain.as(ExitCode.Success)
 }
 
-class Worker[F[_]](number: Int, ref: Ref[F, Int])
-                  (implicit F: Effect[F]) {
+class Worker[F[_]](number: Int, ref: Ref[F, Int])(implicit F: Effect[F]) {
 
-  private val sink: Sink[F, Int] = _.evalMap(n => F.delay(println(s"#$number >> $n")))
+  private val sink: Sink[F, Int] =
+    _.evalMap(n => F.delay(println(s"#$number >> $n")))
 
   def start: Stream[F, Unit] =
     for {
       _ <- Stream.eval(ref.get) to sink
-      _ <- Stream.eval(ref.modify(_ + 1))
+      _ <- Stream.eval(ref.update(_ + 1))
```

Comment:
If I had to rewrite this right now I'd probably remove all the stream machinery and rename that `start` to avoid confusion with the method introduced in `Concurrent`:

```scala
def putStrLn[F[_]: Sync, A](a: A): F[Unit] = Sync[F].delay(println(a))

val exec: F[Unit] =
  ref.get.flatMap(putStrLn) *> ref.update(_ + 1) *>  ref.get.flatMap(putStrLn)
```

Or a `for-comprehension` if you wish.